src/library/searchquery.cpp

```diff
@@ -189,12 +209,39 @@ bool CrateFilterNode::match(const TrackPointer& pTrack) const {
 }
 
 QString CrateFilterNode::toSql() const {
-    return QString("id IN (%1)").arg(CrateStorage::formatQueryForTrackIdsByCrateNameLike(m_crateNameLike));
+    return QString("id IN (%1)").arg(
+            m_pCrateStorage->formatQueryForTrackIdsByCrateNameLike(m_crateNameLike));
+}
+
+
+NoCrateFilterNode::NoCrateFilterNode(const CrateStorage* pCrateStorage)
+    : m_pCrateStorage(pCrateStorage),
+      m_matchInitialized(false) {
+}
+
+bool NoCrateFilterNode::match(const TrackPointer& pTrack) const {
+    if (!m_matchInitialized) {
+        CrateTrackSelectResult crateTracks(
+             m_pCrateStorage->selectAllTracksSorted());
+
+        while (crateTracks.next()) {
+            m_matchingTrackIds.push_back(crateTracks.trackId());
```

Comment:
Track IDs might be pushed multiple times, which is not necessary. The result contains tuples of (track id, crate id) sorted by track id, but we only need distinct track ids here.

We can also think about adding a new query instead of post-filtering the result.